File: exercises/domain-5/block-5.3/solution/solution.py
import asyncio
from dataclasses import dataclass


@dataclass
class TaskContext:
    description: str
    files_modified: list[str]
    includes_db_migration: bool
    is_deployment: bool
    lines_changed: int


@dataclass
class ReviewGate:
    name: str
    trigger: object  # Callable
    message: object  # Callable


class ReviewGatedAgent:
    def __init__(self):
        self.gates: list[ReviewGate] = []
        self.gate_log: list[dict] = []

    def add_gate(self, gate: ReviewGate) -> "ReviewGatedAgent":
        self.gates.append(gate)
        return self

    async def evaluate_gates(self, ctx: TaskContext, get_approval) -> bool:
        triggered = []
        for gate in self.gates:
            did_trigger = gate.trigger(ctx)
            self.gate_log.append({"gate": gate.name, "triggered": did_trigger, "task": ctx.description})
            if did_trigger:
                triggered.append(gate)

        if not triggered:
            return True

        message = "\n\n".join(g.message(ctx) for g in triggered)
        return await get_approval(message)

    async def run_task(self, ctx: TaskContext, task, get_approval) -> None:
        print(f"\nTask: {ctx.description}")
        approved = await self.evaluate_gates(ctx, get_approval)
        if not approved:
            print("  ❌ Task rejected")
            return
        await task()


agent = ReviewGatedAgent()
agent.add_gate(ReviewGate("many-files", lambda ctx: len(ctx.files_modified) > 5, lambda ctx: f"⚠️ Modifies {len(ctx.files_modified)} files."))
agent.add_gate(ReviewGate("db-migration", lambda ctx: ctx.includes_db_migration, lambda ctx: "⚠️ DB schema change."))


async def simulate_approval(message: str) -> bool:
    print(f"  [Review]\n  {message}\n  [Auto-approve: YES]")
    return True


async def main():
    await agent.run_task(
        TaskContext("Fix typo", ["README.md"], False, False, 1),
        lambda: asyncio.sleep(0) or print("  ✓ Fixed"),
        simulate_approval
    )
    await agent.run_task(
        TaskContext("Refactor auth", [f"auth/f{i}.ts" for i in range(8)], False, False, 340),
        lambda: asyncio.sleep(0) or print("  ✓ Refactored"),
        simulate_approval
    )
    print("\nGate log:", agent.gate_log)


asyncio.run(main())

File: exercises/domain-5/block-5.3/solution/solution.ts
interface TaskContext {
  description: string
  filesModified: string[]
  includesDbMigration: boolean
  isDeployment: boolean
  linesChanged: number
}

interface ReviewGate {
  name: string
  trigger: (ctx: TaskContext) => boolean
  message: (ctx: TaskContext) => string
}

class ReviewGatedAgent {
  private gates: ReviewGate[] = []
  gateLog: Array<{ gate: string; triggered: boolean; taskDescription: string }> = []

  addGate(gate: ReviewGate): this {
    this.gates.push(gate)
    return this
  }

  async evaluateGates(ctx: TaskContext, getApproval: (msg: string) => Promise<boolean>): Promise<boolean> {
    const triggered = this.gates.filter(gate => {
      const didTrigger = gate.trigger(ctx)
      this.gateLog.push({ gate: gate.name, triggered: didTrigger, taskDescription: ctx.description })
      return didTrigger
    })

    if (triggered.length === 0) return true

    const reviewMessage = triggered.map(g => g.message(ctx)).join("\n\n")
    const approved = await getApproval(reviewMessage)
    return approved
  }

  async runTask(ctx: TaskContext, task: () => Promise<void>, getApproval: (msg: string) => Promise<boolean>): Promise<void> {
    console.log(`\nTask: ${ctx.description}`)
    const approved = await this.evaluateGates(ctx, getApproval)
    if (!approved) {
      console.log("  ❌ Task rejected — gate not approved")
      return
    }
    await task()
  }
}

const agent = new ReviewGatedAgent()
agent
  .addGate({ name: "many-files", trigger: (ctx) => ctx.filesModified.length > 5, message: (ctx) => `⚠️ Modifies ${ctx.filesModified.length} files.` })
  .addGate({ name: "db-migration", trigger: (ctx) => ctx.includesDbMigration, message: () => "⚠️ DB schema change." })
  .addGate({ name: "deployment", trigger: (ctx) => ctx.isDeployment, message: () => "🚀 Deployment to production." })

async function simulateApproval(message: string): Promise<boolean> {
  console.log(`  [Review]\n  ${message}\n  [Auto-approve: YES]`)
  return true
}

async function main() {
  await agent.runTask(
    { description: "Fix typo", filesModified: ["README.md"], includesDbMigration: false, isDeployment: false, linesChanged: 1 },
    async () => console.log("  ✓ Fixed"),
    simulateApproval
  )
  await agent.runTask(
    { description: "Refactor auth", filesModified: Array.from({length:8},(_,i)=>`auth/f${i}.ts`), includesDbMigration: false, isDeployment: false, linesChanged: 340 },
    async () => console.log("  ✓ Refactored"),
    simulateApproval
  )
  console.log("\nGate log:", agent.gateLog)
}

main().catch(console.error)
